README.md
# react-native-sketch

[![Version](https://img.shields.io/npm/v/react-native-sketch.svg?style=flat-square)](http://npm.im/react-native-sketch)
[![Downloads](https://img.shields.io/npm/dm/react-native-sketch.svg?style=flat-square)](http://npm.im/react-native-sketch)
[![Gitter](https://img.shields.io/badge/chat-on%20gitter-1dce73.svg?style=flat-square)](https://gitter.im/jgrancher/react-native-sketch)
[![MIT License](https://img.shields.io/npm/l/react-native-sketch.svg?style=flat-square)](http://opensource.org/licenses/MIT)

*A react-native component for touch-based drawing with an ability to draw over the images.*

![Screenshots](https://cloud.githubusercontent.com/assets/5517450/15202227/ca865758-183b-11e6-8c4e-41080bc04538.jpg "Disclaimer: This is not my signature ;)")

## Getting started

Use react-native to install and link this component to your project:
```bash
$ react-native install react-native-sketch
```

## Usage

### Solid background

```javascript
import React, { Component } from 'react';
import { StyleSheet, Text, TouchableOpacity, View } from 'react-native';
import Sketch from 'react-native-sketch';

const styles = StyleSheet.create({
  container: {
    padding: 20,    
  },
  instructions: {
    fontSize: 16,
    marginBottom: 20,
    textAlign: 'center',
  },
  sketch: {
    height: 250, // Height needed; Default: 200px
    marginBottom: 20,
    backgroundColor: '#ffffff', // Solid White background. Replacement for the fillColor property
  },
  button: {
    alignItems: 'center',
    backgroundColor: '#111111',
    padding: 20,
  },
  buttonText: {
    color: '#ffffff',
    fontSize: 16,
  },
});

class Signature extends Component {

  constructor(props) {
    super(props);
    this.onReset = this.onReset.bind(this);
    this.onSave = this.onSave.bind(this);
    this.onUpdate = this.onUpdate.bind(this);
  }

  state = {
    encodedSignatures: [],
  };

  /**
   * Do extra things after the sketch reset
   */
  onReset() {
    console.log('The drawing has been cleared!');
  }

  /**
   * The Sketch component provides a 'saveImage' function (promise),
   * so that you can save the drawing in the device and get an object
   * once the promise is resolved, containing the path of the image.
   */
  onSave() {
    this.sketch.saveImage(this.state.encodedSignature)
      .then((data) => console.log(data))
      .catch((error) => console.log(error));
  }

  /**
   * On every update (touch up from the drawing),
   * you'll receive the base64 representation of the drawing as a callback.
   * So we store it in an array of all past lines in order to have an undo functionality
   */
  onUpdate(base64Image) {
    this.setState({ encodedSignatures: [ base64Image, ...this.state.encodedSignatures ] });
  }
  /**
   * Since we stored every past image we can undo through them
   */
  undo() {
        const [ first, ...encodedSignatures ] = this.state.encodedSignatures;

        this.setState( { encodedSignatures } )

        if ( encodedSignatures[ 0 ] ) {

            this.sketch.setImage( encodedSignatures[ 0 ] );

        } else {

            // Just clear the screen
            this.sketch.clear();
        }
    }

  render() {
    return (
      <View style={styles.container}>
        <Text style={styles.instructions}>
          Use your finger on the screen to sign.
        </Text>
        <Sketch
          strokeColor="#111111"
          strokeThickness={2}
          onReset={this.onReset}
          onUpdate={this.onUpdate}
          ref={(sketch) => { this.sketch = sketch; }}
          style={styles.sketch}
        />
        <TouchableOpacity
          disabled={!this.state.encodedSignature}
          style={styles.button}
          onPress={this.onSave}
        >
          <Text style={styles.buttonText}>Save</Text>
        </TouchableOpacity>
      </View>
    );
  }

}

export default Signature;
```

### Render over the image (transparent background) 

1. Set the transparent backgroundColor style property. 

```javascript
const styles = StyleSheet.create({
  sketch: {
    height: 250, 
    backgroundColor: 'rgba(255, 255, 255, 0)', // Transparent background
  }
});
  ```
2. Wrap the Sketch component into the Image component
  
```javascript
  <Image
    style={ styles.imageStyle }
    source={ {uri: imageUri} }
  >
    <Sketch
      strokeColor="#111111"
      strokeThickness={2}
      onReset={this.onReset}
      onUpdate={this.onUpdate}
      ref={(sketch) => { this.sketch = sketch; }}
      style={styles.sketch}
    />
  </Image>
 ```

## Roadmap

- [ ] Define a way for an external component to clear the current drawing.
- [ ] Improve the documentation.
- [ ] Make some tests.
- [ ] Android support (help wanted ¯\\_(ツ)_/¯).

## Notes

This component uses this [smooth freehand drawing technique](http://code.tutsplus.com/tutorials/smooth-freehand-drawing-on-ios--mobile-13164) under the hood.

## Contributing

Feel free to contribute by sending a pull request or [creating an issue](https://github.com/jgrancher/react-native-sketch/issues/new).

## License

MIT


CHANGELOG.md
# Change Log
All notable changes to this project will be documented in this file.
This project adheres to [Semantic Versioning](http://semver.org/).

## [v0.2.5] - 2016-12-06
### Changed
- Replaced a `fillColor` property with a `backgroundColor` style to allow drawing with a transparent background  

## [v0.2.4] - 2016-06-06
### Changed
- Corrected `PropTypes` import in index.ios.js - Removed from 'react-native' package...

## [v0.2.3] - 2016-05-16
### Changed
- Corrected `onReset` callback: Call `onUpdate` with a null value, then the onReset property.

## [v0.2.2] - 2016-05-13
### Changed
- Updated package.json with new keywords

## [v0.2.1] - 2016-05-12
### Changed
- Fix: Replaced a custom UIColor category with RCTConvert (To prevent linking issues...)

## [v0.2.0] - 2016-05-12
### Changed
- The sketch frame is not static anymore (height is 200px by default)
- The clear button is now on the top left corner. (will be hopefully removed in the future)
- Completed README.md with more informations

### Added
- Property `fillColor` (hexa string) to change the background color of the canvas
- Property `strokeColor` (hexa string) to change the stroke color
- Property `strokeThickness` (number) to change the stroke thickness
- Method `saveImage` (async) which resolves with the drawing image path.

## [v0.1.0] - 2016-05-10
### Added
- Initial commit (static sketch frame size & background)


index.ios.js
import React from "react";
import {NativeModules, requireNativeComponent, StyleSheet, View} from "react-native";

const {func, number, string} = React.PropTypes;
const SketchManager = NativeModules.RNSketchManager || {};
const BASE_64_CODE = 'data:image/png;base64,';
const styles = StyleSheet.create({
  base: {
    flex: 1,
    height: 200,
  },
});

export default class Sketch extends React.Component {
  static propTypes = {
    onReset: func,
    onUpdate: func,
    strokeColor: string,
    strokeThickness: number,
    style: View.propTypes.style,
  };

  static defaultProps = {
    onReset: () => {},
    onUpdate: () => {},
    strokeColor: '#000000',
    strokeThickness: 1,
    style: null,
  };

  constructor(props) {
    super(props);
    this.onReset = this.onReset.bind(this);
    this.onUpdate = this.onUpdate.bind(this);
  }

  onReset() {
    this.props.onUpdate(null);
    this.props.onReset();
  }

  onUpdate(e) {
    if (e.nativeEvent.image) {
      this.props.onUpdate(`${BASE_64_CODE}${e.nativeEvent.image}`);
    } else {
      this.onReset();
    }
  }

  saveImage(image) {
    if (typeof image !== 'string') {
      return Promise.reject('You need to provide a valid base64 encoded image.');
    }

    const src = image.indexOf(BASE_64_CODE) === 0 ? image.replace(BASE_64_CODE, '') : image;
    return SketchManager.saveImage(src);
  }

  render() {
    return (
      <RNSketch
        {...this.props}
        onChange={this.onUpdate}
        style={[styles.base, this.props.style]}
      />
    );
  }

}

const RNSketch = requireNativeComponent('RNSketch', Sketch);


RNSketch/RNSketchManager.m
//
//  RNSketchManager.m
//  RNSketch
//
//  Created by Jeremy Grancher on 28/04/2016.
//  Copyright © 2016 Jeremy Grancher. All rights reserved.
//

#import "RNSketchManager.h"
#import "RNSketch.h"
#import "RCTBridge.h"
#import "RCTConvert.h"
#import "RCTEventDispatcher.h"

#define ERROR_IMAGE_INVALID @"ERROR_IMAGE_INVALID"
#define ERROR_FILE_CREATION @"ERROR_FILE_CREATION"

@implementation RNSketchManager

RCT_EXPORT_MODULE()

#pragma mark - Events

RCT_EXPORT_VIEW_PROPERTY(onReset, RCTBubblingEventBlock);
RCT_EXPORT_VIEW_PROPERTY(onChange, RCTBubblingEventBlock);

#pragma mark - Properties

RCT_EXPORT_VIEW_PROPERTY(fillColor, UIColor);
RCT_EXPORT_VIEW_PROPERTY(strokeColor, UIColor);
RCT_EXPORT_VIEW_PROPERTY(clearButtonHidden, BOOL);
RCT_EXPORT_VIEW_PROPERTY(strokeThickness, NSInteger);
RCT_EXPORT_VIEW_PROPERTY(imageType, NSString);

#pragma mark - Lifecycle

- (instancetype)init
{
  if ((self = [super init])) {
    self.sketchView = nil;
  }
  
  return self;
}

- (UIView *)view
{
  if (!self.sketchView) {
    self.sketchView = [[RNSketch alloc] initWithFrame:CGRectZero];
  }
  
  return self.sketchView;
}


#pragma mark - Exported methods

RCT_EXPORT_METHOD(saveImage:(NSString *)encodedImage
                  ofType:(NSString *)imageType
                  resolve:(RCTPromiseResolveBlock)resolve
                  reject:(RCTPromiseRejectBlock)reject)
{
  // Strip the Base 64 Code out if it's there.
  NSString *base64Code = [self.sketchView base64Code];
  encodedImage = [encodedImage stringByReplacingOccurrencesOfString:base64Code
                                                         withString:@""
                                                            options:NULL
                                                              range:NSMakeRange(0, [base64Code length])];
  
  // Create image data with base64 source
  NSData *imageData = [[NSData alloc] initWithBase64EncodedString:encodedImage options:NSDataBase64DecodingIgnoreUnknownCharacters];
  if (!imageData) {
    return reject(ERROR_IMAGE_INVALID, @"You need to provide a valid base64 encoded image.", nil);
  }

  // Create full path of image
  NSArray *paths = NSSearchPathForDirectoriesInDomains(NSDocumentDirectory, NSUserDomainMask, YES);
  NSString *documentsDirectory = [paths firstObject];
  NSFileManager *fileManager = [NSFileManager defaultManager];
  NSString *fullPath = [[documentsDirectory stringByAppendingPathComponent:[[NSUUID UUID] UUIDString]] stringByAppendingPathExtension:imageType];

  // Save image and return the path
  BOOL fileCreated = [fileManager createFileAtPath:fullPath contents:imageData attributes:nil];
  if (!fileCreated) {
    return reject(ERROR_FILE_CREATION, @"An error occured. Impossible to save the file.", nil);
  }
  resolve(@{@"path": fullPath});
}

RCT_EXPORT_METHOD(clear)
{
  dispatch_async(dispatch_get_main_queue(), ^{
    [self.sketchView clearDrawing];
  });
}

@end


RNSketch/RNSketch.m
//
//  RNSketch.m
//  RNSketch
//
//  Created by Jeremy Grancher on 28/04/2016.
//  Copyright © 2016 Jeremy Grancher. All rights reserved.
//

#import "RNSketch.h"
#import "RNSketchManager.h"
#import "RCTEventDispatcher.h"
#import "RCTView.h"
#import "UIView+React.h"

@implementation RNSketch
{
  // Internal
  RCTEventDispatcher *_eventDispatcher;
  UIButton *_clearButton;
  UIBezierPath *_path;
  UIImage *_image;
  CGPoint _points[5];
  uint _counter;

  // Configuration settings
  UIColor *_strokeColor;
}


#pragma mark - UIViewHierarchy methods


- (instancetype)initWithEventDispatcher:(RCTEventDispatcher *)eventDispatcher
{
  if ((self = [super init])) {
    // Internal setup
    self.multipleTouchEnabled = NO;
    _eventDispatcher = eventDispatcher;
    _path = [UIBezierPath bezierPath];

    // TODO: Find a way to get an functionnal external 'clear button'
    [self initClearButton];
  }

  return self;
}

- (void)layoutSubviews
{
  [super layoutSubviews];
  [self drawBitmap];
}


#pragma mark - Subviews


- (void)initClearButton
{
  // Clear button
  CGRect frame = CGRectMake(0, 0, 50, 50);
  _clearButton = [UIButton buttonWithType:UIButtonTypeSystem];
  _clearButton.frame = frame;
  _clearButton.enabled = false;
  _clearButton.tintColor = [UIColor blackColor];
  _clearButton.titleLabel.font = [UIFont boldSystemFontOfSize:18];
  [_clearButton setTitle:@"x" forState:UIControlStateNormal];
  [_clearButton addTarget:self action:@selector(clearDrawing) forControlEvents:UIControlEventTouchUpInside];

  // Clear button background
  UIButton *background = [UIButton buttonWithType:UIButtonTypeCustom];
  background.frame = frame;

  // Add subviews
  [self addSubview:background];
  [self addSubview:_clearButton];
}


#pragma mark - UIResponder methods


- (void)touchesBegan:(NSSet<UITouch *> *)touches withEvent:(UIEvent *)event
{
  _counter = 0;
  UITouch *touch = [touches anyObject];
  _points[0] = [touch locationInView:self];
}

- (void)touchesMoved:(NSSet<UITouch *> *)touches withEvent:(UIEvent *)event
{
  _counter++;
  UITouch *touch = [touches anyObject];
  _points[_counter] = [touch locationInView:self];

  if (_counter == 4) [self drawCurve];
}

- (void)touchesEnded:(NSSet<UITouch *> *)touches withEvent:(UIEvent *)event
{
  // Enabling to clear
  [_clearButton setEnabled:true];

  [self drawBitmap];
  [self setNeedsDisplay];

  [_path removeAllPoints];
  _counter = 0;

  // Send event
  NSDictionary *bodyEvent = @{
                              @"target": self.reactTag,
                              @"image": [self drawingToString],
                              };
  [_eventDispatcher sendInputEventWithName:@"topChange" body:bodyEvent];
}

- (void)touchesCancelled:(NSSet<UITouch *> *)touches withEvent:(UIEvent *)event
{
  [self touchesEnded:touches withEvent:event];
}


#pragma mark - UIViewRendering methods


- (void)drawRect:(CGRect)rect
{
  [_image drawInRect:rect];
  [_strokeColor setStroke];
  [_path stroke];
}


#pragma mark - Drawing methods


- (void)drawCurve
{
  // Move the endpoint to the middle of the line
  _points[3] = CGPointMake((_points[2].x + _points[4].x) / 2, (_points[2].y + _points[4].y) / 2);

  [_path moveToPoint:_points[0]];
  [_path addCurveToPoint:_points[3] controlPoint1:_points[1] controlPoint2:_points[2]];

  [self setNeedsDisplay];

  // Replace points and get ready to handle the next segment
  _points[0] = _points[3];
  _points[1] = _points[4];
  _counter = 1;
}

- (void)drawBitmap
{
  UIGraphicsBeginImageContextWithOptions(self.bounds.size, NO, 0);

  // Draw with context
  [_image drawAtPoint:CGPointZero];
  [_strokeColor setStroke];
  [_path stroke];
  _image = UIGraphicsGetImageFromCurrentImageContext();

  UIGraphicsEndImageContext();
}


#pragma mark - Export drawing


- (NSString *)drawingToString
{
  return [UIImagePNGRepresentation(_image) base64EncodedStringWithOptions:NSDataBase64Encoding64CharacterLineLength];
}


#pragma mark - Clear drawing


- (void)clearDrawing
{
  // Disabling to clear
  [_clearButton setEnabled:false];

  _image = nil;

  [self drawBitmap];
  [self setNeedsDisplay];

  // Send event
  NSDictionary *bodyEvent = @{
                              @"target": self.reactTag,
                              };
  [_eventDispatcher sendInputEventWithName:@"topChange" body:bodyEvent];
}


#pragma mark - Setters

- (void)setStrokeColor:(UIColor *)strokeColor
{
  _strokeColor = strokeColor;
}

- (void)setStrokeThickness:(NSInteger)strokeThickness
{
  _path.lineWidth = strokeThickness;
}

@end
